ext/dom/element.c

```diff
@@ -418,9 +418,68 @@ PHP_METHOD(DOMElement, setAttribute)
 }
 /* }}} end dom_element_set_attribute */
 
-static bool dom_remove_attribute(xmlNodePtr attrp)
+typedef struct {
+	xmlNodePtr current_node;
+	xmlNsPtr defined_ns;
+} dom_deep_ns_redef_item;
+
+/* Reconciliation for a *single* namespace, but reconciles *closest* to the subtree needing it. */
+static void dom_deep_ns_redef(xmlNodePtr node, xmlNsPtr ns_to_redefine)
 {
+	size_t worklist_capacity = 128;
+	dom_deep_ns_redef_item *worklist = emalloc(sizeof(dom_deep_ns_redef_item) * worklist_capacity);
+	worklist[0].current_node = node;
+	worklist[0].defined_ns = NULL;
+	size_t worklist_size = 1;
+
+	while (worklist_size > 0) {
+		worklist_size--;
+		dom_deep_ns_redef_item *current_worklist_item = &worklist[worklist_size];
+		ZEND_ASSERT(current_worklist_item->current_node->type == XML_ELEMENT_NODE);
+		xmlNsPtr defined_ns = current_worklist_item->defined_ns;
+
+		if (current_worklist_item->current_node->ns == ns_to_redefine) {
+			if (defined_ns == NULL) {
+				defined_ns = xmlNewNs(current_worklist_item->current_node, ns_to_redefine->href, ns_to_redefine->prefix);
+			}
+			current_worklist_item->current_node->ns = defined_ns;
+		}
+
+		for (xmlAttrPtr attr = current_worklist_item->current_node->properties; attr; attr = attr->next) {
+			if (attr->ns == ns_to_redefine) {
+				if (defined_ns == NULL) {
+					defined_ns = xmlNewNs(current_worklist_item->current_node, ns_to_redefine->href, ns_to_redefine->prefix);
+				}
+				attr->ns = defined_ns;
+			}
+		}
+
+		for (xmlNodePtr child = current_worklist_item->current_node->children; child; child = child->next) {
+			if (child->type != XML_ELEMENT_NODE) {
+				continue;
+			}
+			if (worklist_size == worklist_capacity) {
+				if (UNEXPECTED(worklist_capacity >= SIZE_MAX / 3 * 2 / sizeof(dom_deep_ns_redef_item))) {
+					/* Shouldn't be possible to hit, but checked for safety anyway */
+					return;
+				}
+				worklist_capacity = worklist_capacity * 3 / 2;
+				worklist = erealloc(worklist, sizeof(dom_deep_ns_redef_item) * worklist_capacity);
```

Comment:
> This looks sensible to me, maybe this should land in 8.3?

Probably. Could be landed in even lower versions as this is a proper bugfix, but I don't dare to do that because of reasons you know...

> Maybe use safe_erealloc() instead?

I have to check for overflow of doing `*3/2` already, so I folded the overflow check for `sizeof(dom_deep_ns_redef_item) * worklist_capacity` into that one already. I therefore don't think it's necessary to use `safe_erealloc`.

Reply:
ACK for the realloc.

8.3 _should_ be safe as it hasn't been "properly" released yet IMHO.